xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -471,70 +429,80 @@ public Result selectConfig(PickSubchannelArgs args) {
+                  committed = true;
+                  releaseCluster(finalCluster);
+                  delegate().onHeaders(headers);
+                }
 
-                        @Override
-                        public void onClose(Status status, Metadata trailers) {
-                          if (!committed) {
-                            releaseCluster(finalCluster);
-                          }
-                          delegate().onClose(status, trailers);
-                        }
-                      };
-                      delegate().start(listener, headers);
-                    }
-                  };
+                @Override
+                public void onClose(Status status, Metadata trailers) {
+                  if (!committed) {
+                    releaseCluster(finalCluster);
+                  }
+                  delegate().onClose(status, trailers);
                 }
               };
+              delegate().start(listener, headers);
+            }
+          };
+        }
+      }
 
-          Executor callExecutor = callOptions.getExecutor();
-          if (callExecutor == null) { // This should never happen in practice because
-            // ManagedChannelImpl.ConfigSelectingClientCall always provides CallOptions with
-            // a callExecutor.
-            // TODO(https://github.com/grpc/grpc-java/issues/7868)
-            callExecutor = MoreExecutors.directExecutor();
-          }
-          if (finalDelayNanos != null && finalAbortStatus != null) {
-            return new ActiveFaultCountingClientCall<>(
-                new DelayInjectedCall<>(
-                    finalDelayNanos, callExecutor, scheduler, callOptionsForCluster.getDeadline(),
-                    Suppliers.ofInstance(
-                        new FailingClientCall<ReqT, RespT>(finalAbortStatus, callExecutor))));
-          }
-          if (finalAbortStatus != null) {
-            return new ActiveFaultCountingClientCall<>(
-                new FailingClientCall<ReqT, RespT>(finalAbortStatus, callExecutor));
-          }
-          if (finalDelayNanos != null) {
-            return new ActiveFaultCountingClientCall<>(
-                new DelayInjectedCall<>(
-                    finalDelayNanos, callExecutor, scheduler, callOptionsForCluster.getDeadline(),
-                    configApplyingCallSupplier));
+      filterInterceptors.add(new ClusterSelectionInterceptor());
+      return
+          Result.newBuilder()
+              .setConfig(config)
+              .setInterceptor(combineInterceptors(filterInterceptors))
+              .build();
+    }
+
+    // Adds LameFilter to the end of filter chain due to absence of RouterFilter.
+    private Result lameResult(RoutingConfig savedRoutingConfig, PickSubchannelArgs args) {
+      Map<String, ?> rawServiceConfig = Collections.emptyMap();
+      if (enableTimeout) {
+        Long timeoutNano = routingConfig.fallbackTimeoutNano;
```

Comment:
Again, are we able to reorganize the code so that we do not need to repeat this logic?

Reply:
Done.